core/services/pipeline/orm.go

```diff
@@ -0,0 +1,360 @@
+                          AND pipeline_task_runs.pipeline_run_id = predecessor_unfinished_runs.pipeline_run_id
+                WHERE pipeline_task_runs.finished_at IS NULL
+                GROUP BY (pipeline_task_runs.id)
+                HAVING (
+                    bool_and(predecessor_unfinished_runs.finished_at IS NOT NULL)
+                    OR
+                    count(predecessor_unfinished_runs.id) = 0
+                )
+            )
+            LIMIT 1
+            FOR UPDATE OF pipeline_task_runs SKIP LOCKED;
+        `).Scan(&ptRun).Error
+		if gorm.IsRecordNotFoundError(err) {
+			done = true
+			return nil
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run")
+		}
+
+		// Fill in the TaskSpec
+		err = tx.Where("id = ?", ptRun.PipelineTaskSpecID).First(&ptRun.PipelineTaskSpec).Error
+		if gorm.IsRecordNotFoundError(err) {
+			// done = true
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run's spec")
+		}
+
+		// Fill in the PipelineRun
+		err = tx.Where("id = ?", ptRun.PipelineRunID).First(&ptRun.PipelineRun).Error
+		if gorm.IsRecordNotFoundError(err) {
+			// done = true
+			return err
+		} else if err != nil {
+			return errors.Wrap(err, "error finding next task run's pipeline.Run")
+		}
+
+		// Find all the predecessor task runs
+		err = tx.Raw(`
+                SELECT pipeline_task_runs.* FROM pipeline_task_runs
+                LEFT JOIN pipeline_task_specs AS specs_right ON specs_right.id = pipeline_task_runs.pipeline_task_spec_id
+                LEFT JOIN pipeline_task_specs AS specs_left ON specs_left.id = specs_right.successor_id
+                LEFT JOIN pipeline_task_runs AS successors ON successors.pipeline_task_spec_id = specs_left.id
+                      AND successors.pipeline_run_id = pipeline_task_runs.pipeline_run_id
+                WHERE successors.id = ?`, ptRun.ID).Find(&predecessors).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding task run predecessors")
+		}
+
+		// Get the job ID
+		var job struct{ ID int32 }
+		err = tx.Raw(`
+            SELECT jobs.id FROM pipeline_task_runs
+            LEFT JOIN pipeline_task_specs ON pipeline_task_specs.id = pipeline_task_runs.pipeline_task_spec_id
+            LEFT JOIN jobs ON jobs.pipeline_spec_id = pipeline_task_specs.pipeline_spec_id
+            WHERE pipeline_task_runs.id = ?
+        `, ptRun.ID).Scan(&job).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding job ID")
+		}
+
+		// Call the callback
+		result := fn(job.ID, ptRun, predecessors)
+
+		// Update the task run record with the output and error
+		if result.Value != nil {
+			out := &JSONSerializable{Val: result.Value}
+			err = tx.Exec(`UPDATE pipeline_task_runs SET output = ?, error = NULL, finished_at = ? WHERE id = ?`, out, time.Now(), ptRun.ID).Error
```

Comment:
Looks like we can pull the insert out of this if and just populate the values for it, no?

Reply:
I don't follow, which insert?  And what do you have in mind when you say "populate the value for it"?